components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -137,14 +147,29 @@ open class GleanInternalAPI internal constructor () {
      *
      * Metric collection is enabled by default.
      *
-     * When disabled, metrics aren't recorded at all and no data
+     * When uploading is disabled, metrics aren't recorded at all and no data
      * is uploaded.
      *
+     * When disabling, all pending metrics, events and queued pings are cleared.
+     *
+     * When enabling, the core Glean metrics are recreated.
+     *
      * @param enabled When true, enable metric collection.
+     * @param applicationContext The application context.
      */
     fun setUploadEnabled(enabled: Boolean) {
         logger.info("Metrics enabled: $enabled")
+        val origUploadEnabled = uploadEnabled
+        // Change the value of uploadEnabled before initializeCoreMetrics so
+        // that the core metrics are actually recorded in the stores.
         uploadEnabled = enabled
+        if (isInitialized()) {
+            if (origUploadEnabled == true && enabled == false) {
+                clearMetrics()
```

Comment:
@mdboom Mh, hang on. I just realized we might inadvertently delete all the stored data when not wanted. We encourage calling `setUploadEnabled` *before* the init. Could this happen to remove all the local data (including the client id) if things are flipped at that time?

Reply:
If we call `setUploadEnabled` before `init`, nothing happens other than setting the flag.  This is in part because if it goes from `false` to `true`, it would try to regenerate the application metrics, but since Glean is uninitialized it wouldn't do anything.  And in general, this guard is there because doing anything too heavy with Glean before it's initialized is probably undefined behavior and full of NPEs etc.  So nothing will ever happen on startup if (as we encourage) `setUploadEnabled` is called before `initialize`.

I think there's potential for confusion here, though, since the enabled state isn't persisted.  It always starts at `true` before calling `setUploadEnabled`, but it might actually have already been set to `false` in a previous run.  I'm trying to work through whether that really matters in practice -- I think not...

Reply:
This has been updated so that `initialize` always handles the state switch once upon startup, and added a couple of tests around startup.  I think it's sorted now so we should always be in a correct state.